woodwork/statistics_utils/_get_dependence_dict.py

```diff
@@ -0,0 +1,201 @@
+    # cut off data if necessary
+    if _is_dask_dataframe(data):
+        data = data.compute()
+    elif _is_koalas_dataframe(dataframe):
+        data = data.to_pandas()
+    if nrows is not None and nrows < data.shape[0]:
+        data = data.sample(nrows, random_state=random_seed)
+
+    notna_mask = data.notnull()
+    not_null_cols = data.columns[notna_mask.any()]
+    if set(not_null_cols) != set(valid_columns):
+        data = data.loc[:, not_null_cols]
+
+    # Setup for progress callback and make initial call
+    # Assume 1 unit for preprocessing, 2n for make categorical and (n*n+n)/2 for main calculation loop
+    n = len(data.columns)
+    total_loops = 1 + 2 * n + (n * n + n) / 2
+    current_progress = _update_progress(
+        start_time, timer(), 1, 0, total_loops, unit, callback
+    )
+
+    data = _bin_numeric_cols_into_categories(dataframe.ww.schema, data, num_bins)
+    current_progress = _update_progress(
+        start_time, timer(), 2 * n, current_progress, total_loops, unit, callback
+    )
+
+    # calculate mutual info for all pairs of columns
+    dependence_list = []
+    col_names = data.columns.to_list()
+    for i, a_col in enumerate(col_names):
+        for j in range(i, len(col_names)):
+            b_col = col_names[j]
+            if a_col == b_col:
+                # Ignore because the mutual info for a column with itself will always be 1
+                current_progress = _update_progress(
+                    start_time,
+                    timer(),
+                    1,
+                    current_progress,
+                    total_loops,
+                    unit,
+                    callback,
+                )
+                continue
+            else:
+                result = {"column_1": a_col, "column_2": b_col}
+                num_intersect = (notna_mask[a_col] & notna_mask[b_col]).sum()
+                too_sparse = num_intersect < min_shared
+                if too_sparse:
+                    # TODO: reword since Pearson can return NaN naturally
+                    warnings.warn(
+                        "One or more values in the returned matrix are NaN. A "
+                        "NaN value indicates there were not enough rows where "
+                        "both columns had non-null data",
+                        SparseDataWarning,
+                    )
+                    result.update({measure_type: np.nan for measure_type in measure})
+                else:
+                    num_union = (notna_mask[a_col] | notna_mask[b_col]).sum()
+                    if calc_mutual:
+                        mi_score = adjusted_mutual_info_score(data[a_col], data[b_col])
+                        mi_score = mi_score * num_intersect / num_union
+                        if "mutual" in measure:
+                            result["mutual"] = mi_score
+                    if calc_pearson:
+                        pearson_score = np.corrcoef(data[a_col], data[b_col])[0, 1]
```

Comment:
I'm not too familiar with how these metrics work, so this may be a dumb question, but do we want to use the binned numeric data for Pearson?